### PR TITLE
Mark subclasses of Dimension final.

### PR DESCRIPTION
**Pull request by tkremenek**
Possible resolution of blocking adoption of `-swift-version 5`:

https://github.com/apple/swift/pull/20374

### CODE REVIEW

**Comment by tkremenek** [comment]
@swift-ci test

File: Foundation/Unit.swift [hunk 1]
```diff
@@ -248,7 +248,7 @@ open class Dimension : Unit {
     }
 }
 
-open class UnitAcceleration : Dimension {
+final class UnitAcceleration : Dimension {
```
Comment (millenomi) [inline Foundation/Unit.swift]: I don't think we can take this patch? These subclasses are open in ObjC, and we want as-close-as-possible source parity.

Comment (millenomi) [reply Foundation/Unit.swift]: (I assume we're trying to work around the fact that Swift 5 prevents us from having a pure Swift class that has the equivalent of `+ (instancetype)baseUnit`?)

Comment (tkremenek) [reply Foundation/Unit.swift]: At this point I'm testing the option for consideration.  If we took it there are other things to consider, such as parity with Objective-C and expectations for users of the APIs.

Comment (millenomi) [reply Foundation/Unit.swift]: IIRC, subclassing is part of the contract of this API. (I will check tomorrow.)

I have a patch I can PR that moves .baseUnit() to be of Dimension type everywhere, erasing the type but allowing the build to work. I will discuss it with @parkera tomorrow morning.

**Comment by tkremenek** [comment]
@swift-ci test

File: Foundation/Unit.swift (hunk 1 above)
Comment (millenomi) [reply Foundation/Unit.swift]: Yeah, and Swift is just highlighting it. :(

I'll circle back before EOD.

Comment (millenomi) [reply Foundation/Unit.swift]: (It feels like Swift could allow `Self` class methods if they were similar to `required` initializers, wherein it is not valid to express a class that doesn't override the method.)

Comment (tkremenek) [reply Foundation/Unit.swift]: Yes, that sounds like potentially a reasonable language extension that does not exist today.

Comment (parkera) [reply Foundation/Unit.swift]: For now, to unblock Swift 5 adoption, we will take this patch.

**Comment by millenomi** [comment]
We had a side discussion between me, @tkremenek and @parkera and decided that the right way forward is to require subclasses to be made `final`.

**Comment by millenomi** [comment]
@swift-ci please test and merge

**Pull request merged**